File: app/src/main/java/com/grsu/guideapp/fragments/categories/CatalogFragment.java
package com.grsu.guideapp.fragments.categories;

import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.grsu.guideapp.R;
import com.grsu.guideapp.adapters.TypesAdapter;
import com.grsu.guideapp.base.BaseFragment;
import com.grsu.guideapp.base.listeners.ItemClickListener;
import com.grsu.guideapp.database.Test;
import com.grsu.guideapp.fragments.categories.CategoriesContract.CategoriesViews;
import com.grsu.guideapp.fragments.list_objects.ListObjectFragment;
import com.grsu.guideapp.models.DtoType;
import java.util.List;

public class CatalogFragment extends BaseFragment<CategoriesPresenter, FragmentActivity>
        implements CategoriesViews, ItemClickListener {

    private List<DtoType> mTypes;
    TypesAdapter adapter;
//    @BindView(R.id.fragment_catalog_rv)
    RecyclerView fragment_catalog_rv;
//    @BindView(R.id.fragment_catalog_empty_tv)
    TextView fragment_catalog_empty_tv;

    @NonNull
    @Override
    protected CategoriesPresenter getPresenterInstance() {
        return new CategoriesPresenter(this, new CategoriesInteractor(new Test(getContext())));
    }

    @Override
    protected int getLayout() {
        return R.layout.fragment_catalog;
    }

    @Override
    protected String getTitle() {
        return getString(R.string.list_objects_fragment);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        setHasOptionsMenu(true);

//        getActivity.setTitleToolbar(getTitle());
        fragment_catalog_rv.setHasFixedSize(true);
        fragment_catalog_rv.setLayoutManager(new LinearLayoutManager(getContext()));
        fragment_catalog_rv.addItemDecoration(new DividerItemDecoration(getContext(), DividerItemDecoration.VERTICAL));
        adapter = new TypesAdapter(null, this);
        fragment_catalog_rv.setAdapter(adapter);
        mPresenter.getAllTypes(getString(R.string.locale));
        return view;
    }

    @Override
    public void onItemClick(View view, int position) {
        DtoType type = mTypes.get(position);
        ListObjectFragment fragment = ListObjectFragment.newInstance(type.getId(), type.getName());
        getActivity.getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.container, fragment)
                .addToBackStack(fragment.getClass().getName() + "_" + position)
                .commit();
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.list_routes, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (R.id.action_list_routes == item.getItemId()) {
//            getActivity.openListRoutesFragment();
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void updateList(List<DtoType> types) {
        mTypes = types;
        if (adapter != null) {
            adapter.setTypeList(types);
        }
    }

    @Override
    public void emptyData() {
        fragment_catalog_rv.setVisibility(View.GONE);
        fragment_catalog_empty_tv.setVisibility(View.VISIBLE);
        fragment_catalog_empty_tv.setText(R.string.empty_list);
    }
}

File: app/src/main/java/com/grsu/guideapp/fragments/catalog/CatalogFragment.kt
package com.grsu.guideapp.fragments.catalog

import android.os.Bundle
import android.view.*
import android.widget.ProgressBar
import android.widget.TextView
import androidx.activity.addCallback
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.grsu.guideapp.R
import com.grsu.guideapp.activities.SharedViewModel
import com.grsu.guideapp.adapters.TypesAdapter
import com.grsu.guideapp.base.Result
import com.grsu.guideapp.database.vo.TypeItemVO
import com.grsu.guideapp.utils.extensions.hide
import com.grsu.guideapp.utils.extensions.navigate
import com.grsu.guideapp.utils.extensions.show
import timber.log.Timber

class CatalogFragment : Fragment(), ((View, Int) -> Unit), Observer<Result<List<TypeItemVO>>> {

    private var typesAdapter: TypesAdapter? = null
    private lateinit var recyclerView: RecyclerView
    private lateinit var textError: TextView
    private lateinit var progress: ProgressBar
    private lateinit var model: CatalogViewModel
    private lateinit var sharedModel: SharedViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        typesAdapter = TypesAdapter(listener = this)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(R.layout.fragment_catalog, container, false)
        setHasOptionsMenu(true)


        model = ViewModelProvider(this)[CatalogViewModel::class.java]
        sharedModel = ViewModelProvider(requireActivity())[SharedViewModel::class.java]
        requireActivity().onBackPressedDispatcher.addCallback(viewLifecycleOwner) {
            sharedModel.finishApplication()
        }

        recyclerView = view.findViewById(R.id.rv_fragment_catalog)
        textError = view.findViewById(R.id.tv_fragment_catalog_text_error)
        progress = view.findViewById(R.id.pb_fragment_catalog_progress)

        with(recyclerView) {
            setHasFixedSize(true)
            val manager = LinearLayoutManager(requireContext())
            layoutManager = manager
            addItemDecoration(DividerItemDecoration(requireContext(), manager.orientation))
            adapter = typesAdapter
        }

        model.getData().observe(viewLifecycleOwner, this)
        return view
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.list_routes, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (R.id.action_list_routes == item.itemId) {
            navigate(R.id.action_catalog_to_routes)
        }

        return super.onOptionsItemSelected(item)
    }

    override fun invoke(view: View, position: Int) {
        typesAdapter?.getItem(position)?.also {
            navigate(CatalogFragmentDirections.actionCatalogToObjects(it.id.toInt(), it.name))
        }
        Timber.e("Click item position = $position")
    }

    override fun onChanged(result: Result<List<TypeItemVO>>?) {
        when (result) {
            is Result.Success -> {
                typesAdapter?.submitList(result.data)
                recyclerView.show()
                textError.hide()
                progress.hide()
            }
            is Result.Loading -> {
                recyclerView.hide()
                textError.hide()
                progress.show()
            }
            is Result.Error -> {
                recyclerView.hide()
                textError.show()
                progress.hide()
                textError.text = result.error
            }
        }
    }

    override fun onDestroyView() {
        recyclerView.adapter = null
        super.onDestroyView()
    }
}